README.md
# Chargehound python bindings 

[![Build Status](https://travis-ci.org/chargehound/chargehound-python.svg?branch=master)](https://travis-ci.org/chargehound/chargehound-python) [![PyPI version](https://badge.fury.io/py/chargehound.svg)](https://badge.fury.io/py/chargehound)

## Installation

`pip install chargehound`

## Usage

Import chargehound and set your API key.

```python
import chargehound
chargehound.api_key = '{ YOUR_API_KEY }'
```

### Requests

Every resource is accessed via the Chargehound module.

```python
dispute = chargehound.Disputes.submit('dp_123', fields={'customer_name': 'Susie'})
```

### Response

Responses from the API are automatically parsed from JSON and returned as Python objects.

Responses also include the HTTP status code on the response object as the status field.

```python
dispute = chargehound.Disputes.retrieve('dp_123')

print dispute.state
# 'needs_response'
print dispute.response.status
# 200
```

## Documentation

[Disputes](https://www.chargehound.com/docs/api/index.html?python#disputes)

[Errors](https://www.chargehound.com/docs/api/index.html?python#errors)

## Development

To build and install from the latest source:

```bash
$ git clone git@github.com:chargehound/chargehound-python.git
$ pip install -r dev_requirements.txt
```

Run tests:

```bash
$ python setup.py test
```

## Deployment

If you added a new depency, make sure the change is reflected in the `requirements.txt` file and
the `setup.py` file.

To deploy a new version of the SDK, you will need Pandoc (http://pandoc.org/) installed.
Pandoc will convert the README.md into the .rst format required for the Python repository.
Instructions are here: [http://pandoc.org/installing.html](http://pandoc.org/installing.html).

Next, install PyPandoc, the Python wrapper for Pandoc, following [these instructions](https://pypi.python.org/pypi/pypandoc).
 
The last pre-requisite is [Twine](https://pypi.python.org/pypi/twine), a utility for interacting with PyPi.

Once Pandoc, PyPandoc and Twine are installed, you can build and deploy a new module to PyPi with the following steps:

 1. Update the CHANGELOG to describe what feature have been added.
 2. Bump the version number in `chargehound/version.py`
 3. Rebuild and deploy the package with:
   ```python setup.py sdist```
 4. Upload the distributable to PyPi using:   
   ```twine upload dist/{name_of_generated_package}```
 4. Confirm the new package version is available at [https://pypi.python.org/pypi/chargehound](https://pypi.python.org/pypi/chargehound)


CHANGELOG
1.0.0 April, 2016
  - Initial release with disputes API
1.1.0 June, 2016
  - Added APIs to pass product lists
1.1.1 July, 2016
  - Included README in .rst format
1.2.0 July, 2016
  - Added timeout option
2.0.0 October, 2016
  - Remove callback option for requests
  - Add Dispute create method
  - Add Dispute response retrieve method
  - Expose response status codes
  - Add typed response models
2.1.0 May, 2017
  - Replace NeoBunch library with Munch library. See https://pypi.python.org/pypi/munch/
2.1.1 May, 2017
  - Add munch to requires
2.1.2 May, 2017
  - Do not import from chargehound before installing dependencies
2.2.0 July, 2017
  - Add Accept endpoint
2.3.0 November, 2017
  - Add support for overriding version
2.3.1 September, 2018
  - Update docs
2.3.2 October, 2018
  - Upgrade requests dependency
2.4.0 Sept, 2019
  - Add tests for past payments
  - Allow setting of host port and protocol
2.5.0 Feb, 2021
  - Surface error types


test/test_error.py
import chargehound
import requests_mock
import unittest2
from chargehound.error import ChargehoundBadRequestError, ChargehoundError


class ErrorTest(unittest2.TestCase):

    def setUp(self):
        super(ErrorTest, self).setUp()
        chargehound.api_key = 'API_KEY'

    @requests_mock.mock()
    def test_bad_request(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/submit',
                  status_code=400,
                  json={
                    'error': {'status': 400, 'message': 'Bad request.'}
                  })

        try:
            chargehound.Disputes.submit('dp_123')
        except ChargehoundBadRequestError as bad:
            assert bad.status == 400
            assert bad.message == 'Bad request.'

    @requests_mock.mock()
    def test_invalid_request_type(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/submit',
                  status_code=400,
                  json={
                    'error': {'status': 400, 'message': 'Bad request.', 'type': 'invalid_request'}
                  })

        try:
            chargehound.Disputes.submit('dp_123')
        except ChargehoundBadRequestError as bad:
            assert bad.status == 400
            assert bad.message == 'Bad request.'
            assert bad.type == 'invalid_request'

    def test_propagate_errors(self):
        orig_host = chargehound.host
        chargehound.host = 'test'

        try:
            chargehound.Disputes.retrieve('dp_123')
        except Exception as e:
            assert not isinstance(e, ChargehoundError)
            chargehound.host = orig_host


test/test_disputes.py
import chargehound
import requests_mock
import unittest2
import json

from chargehound.version import VERSION
from chargehound.models import List, Dispute, Response


get_headers = {
    'accept': 'application/json',
    'authorization': 'Basic QVBJX0tFWTo=',
    'user-agent': 'Chargehound/v1 PythonBindings/' + VERSION
}

post_headers = {
    'accept': 'application/json',
    'authorization': 'Basic QVBJX0tFWTo=',
    'content-type': 'application/json',
    'user-agent': 'Chargehound/v1 PythonBindings/' + VERSION
}

product_info = [{'name': 'Saxophone',
                 'description': 'Alto saxophone, with carrying case',
                 'image': 'http://s3.amazonaws.com/chargehound/saxophone.png',
                 'sku': '17283001272',
                 'quantity': 1,
                 'amount': 20000,
                 'url': 'http://www.example.com', },
                {'name': 'Milk',
                 'description': 'Semi-skimmed Organic',
                 'image': 'http://s3.amazonaws.com/chargehound/milk.png',
                 'sku': '26377382910',
                 'quantity': '64oz',
                 'amount': 400,
                 'url': 'http://www.example.com'}]

correspondence_info = [{'to': 'customer@example.com',
                        'from': 'noreply@example.com',
                        'subject': 'Your Order',
                        'body': 'Your order was received.',
                        'caption': 'Order confirmation email.'},
                       {'to': 'customer@example.com',
                        'from': 'noreply@example.com',
                        'subject': 'Your Order',
                        'body': 'Your order was delivered.',
                        'caption': 'Delivery confirmation email.'}]

past_payments_info = [{'id': 'ch_1',
                        'amount': 20000,
                        'currency': 'usd',
                        'charged_at': '2019-09-10 11:09:41PM UTC'},
                      {'id': 'ch_2',
                        'amount': 50000,
                        'currency': 'usd',
                        'charged_at': '2019-09-03 11:09:41PM UTC'}]

dispute_response = {
  'id': 'dp_123',
  'fields': {'customer_name': 'Susie'},
  'products': [],
  'correspondence': [],
  'past_payments': [],
  'object': 'dispute'
}

dispute_products_response = {
  'id': 'dp_123',
  'object': 'dispute',
  'fields': {'customer_name': 'Susie'},
  'products': product_info,
  'correspondence': [],
  'past_payments': [],
}

dispute_correspondence_response = {
  'id': 'dp_123',
  'object': 'dispute',
  'fields': {'customer_name': 'Susie'},
  'correspondence': correspondence_info,
  'products': [],
  'past_payments': [],
}

dispute_past_payments_response = {
  'id': 'dp_123',
  'object': 'dispute',
  'fields': {'customer_name': 'Susie'},
  'past_payments': past_payments_info,
  'correspondence': [],
  'products': [],
}

dispute_list_response = {
  'object': 'list',
  'data': [{
    'id': 'dp_123',
    'fields': {'customer_name': 'Susie'},
    'products': [],
    'correspondence': [],
    'past_payments': [],
    'object': 'dispute'
  }]
}

response_response = {
  'dispute_id': 'dp_123',
  'object': 'response'
}


def is_json(response_body):
    """Can this HTTP response be parsed as JSON?"""
    try:
        json.loads(response_body)
        return True
    except:
        return False


def json_has_structure(response_body, expected_json):
    """Does the JSON this HTTP response match the expected structure?"""
    actual_json = json.loads(response_body)

    return actual_json == expected_json


class DisputeTest(unittest2.TestCase):

    def setUp(self):
        super(DisputeTest, self).setUp()
        chargehound.api_key = 'API_KEY'

    @requests_mock.mock()
    def test_create_dispute(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes',
                  status_code=200,
                  request_headers=post_headers,
                  json=dispute_response)
        chargehound.Disputes.create({'id': 'dp_123'})
        assert mock.called

    @requests_mock.mock()
    def test_retrieve_dispute(self, mock):
        mock.get('https://api.chargehound.com/v1/disputes/dp_123',
                 status_code=200,
                 request_headers=get_headers,
                 json=dispute_response)
        chargehound.Disputes.retrieve('dp_123')
        assert mock.called

    @requests_mock.mock()
    def test_retrieve_dispute_response(self, mock):
        mock.get('https://api.chargehound.com/v1/disputes/dp_123/response',
                 status_code=200,
                 request_headers=get_headers,
                 json=response_response)
        res = chargehound.Disputes.response('dp_123')

        assert mock.called
        assert isinstance(res, Response)
        assert res.dispute_id == 'dp_123'

    @requests_mock.mock()
    def test_list_all_disputes(self, mock):
        mock.get('https://api.chargehound.com/v1/disputes',
                 status_code=200,
                 request_headers=get_headers,
                 json=dispute_list_response)
        chargehound.Disputes.list()
        assert mock.called

    @requests_mock.mock()
    def test_list_disputes_filter_state(self, mock):
        url = 'https://api.chargehound.com/v1/disputes?state=needs_response'
        mock.get(url,
                 status_code=200,
                 request_headers=get_headers,
                 json=dispute_list_response)
        chargehound.Disputes.list(state='needs_response')
        assert mock.called

    @requests_mock.mock()
    def test_list_disputes_filter_multiple_states(self, mock):
        url = 'https://api.chargehound.com/v1/disputes' +\
            '?state=needs_response&state=warning_needs_response'
        mock.get(url,
                 status_code=200,
                 request_headers=get_headers,
                 json=dispute_list_response)
        chargehound.Disputes.list(
          state=['needs_response', 'warning_needs_response']
        )
        assert mock.called

    @requests_mock.mock()
    def test_submit_dispute(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/submit',
                  status_code=201,
                  json=dispute_response)
        chargehound.Disputes.submit('dp_123',
                                    fields={'customer_name': 'Susie'})

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'}})

    @requests_mock.mock()
    def test_submit_dispute_with_product_info(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/submit',
                  status_code=201,
                  json=dispute_products_response)
        chargehound.Disputes.submit('dp_123',
                                    fields={'customer_name': 'Susie'},
                                    products=product_info)

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'},
                                   'products': product_info})

    @requests_mock.mock()
    def test_submit_dispute_with_correspondence_info(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/submit',
                  status_code=201,
                  json=dispute_correspondence_response)
        chargehound.Disputes.submit('dp_123',
                                    fields={'customer_name': 'Susie'},
                                    correspondence=correspondence_info)

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'},
                                   'correspondence': correspondence_info})

    @requests_mock.mock()
    def test_submit_dispute_with_past_payments_info(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/submit',
                  status_code=201,
                  json=dispute_past_payments_response)
        chargehound.Disputes.submit('dp_123',
                                    fields={'customer_name': 'Susie'},
                                    past_payments=past_payments_info)

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'},
                                   'past_payments': past_payments_info})

    @requests_mock.mock()
    def test_update_dispute(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123',
                  status_code=200,
                  json=dispute_response)
        chargehound.Disputes.update('dp_123',
                                    fields={'customer_name': 'Susie'})

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'}})

    @requests_mock.mock()
    def test_update_dispute_with_product_info(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123',
                  status_code=200,
                  json=dispute_products_response)
        chargehound.Disputes.update('dp_123',
                                    fields={'customer_name': 'Susie'},
                                    products=product_info)

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'},
                                   'products': product_info})

    @requests_mock.mock()
    def test_update_dispute_with_correspondence_info(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123',
                  status_code=200,
                  json=dispute_correspondence_response)
        chargehound.Disputes.update('dp_123',
                                    fields={'customer_name': 'Susie'},
                                    correspondence=correspondence_info)

        assert mock.called

        response = mock.request_history[0].body
        try:
            response = response.decode()
        except AttributeError:
            pass

        assert response
        assert is_json(response)
        assert json_has_structure(response,
                                  {'fields': {'customer_name': 'Susie'},
                                   'correspondence': correspondence_info})

    @requests_mock.mock()
    def test_typed_responses(self, mock):
        mock.get('https://api.chargehound.com/v1/disputes',
                 status_code=200,
                 request_headers=get_headers,
                 json=dispute_list_response)
        li = chargehound.Disputes.list()
        assert mock.called
        assert isinstance(li, List)
        assert isinstance(li.data[0], Dispute)

    @requests_mock.mock()
    def test_typed_responses_can_be_jsonified(self, mock):
        mock.get('https://api.chargehound.com/v1/disputes',
                 status_code=200,
                 request_headers=get_headers,
                 json=dispute_list_response)
        li = chargehound.Disputes.list()

        response = {
          'object': 'list',
          'data': [{
            'id': 'dp_123',
            'fields': {'customer_name': 'Susie'},
            'object': 'dispute',
            'products': [],
            'past_payments': [],
            'correspondence': []
          }],
          'response': [200]
        }

        assert mock.called
        assert json.dumps(response, sort_keys=True) \
            == json.dumps(li, sort_keys=True)

    @requests_mock.mock()
    def test_accept_dispute(self, mock):
        mock.post('https://api.chargehound.com/v1/disputes/dp_123/accept',
                  status_code=200,
                  json=dispute_response)
        chargehound.Disputes.accept('dp_123')

        assert mock.called


test/test_requestor.py
import chargehound
import requests_mock
import unittest2


class ApiRequestorTest(unittest2.TestCase):

    def setUp(self):
        super(ApiRequestorTest, self).setUp()
        chargehound.api_key = 'API_KEY'

    @requests_mock.mock()
    def test_response_status(self, mock):
        mock.get('https://api.chargehound.com/v1/disputes/dp_123',
                 json={'id': 'dp_123'})
        dispute = chargehound.Disputes.retrieve('dp_123')
        assert dispute['response']['status'] == 200


chargehound/error.py
class ChargehoundError(Exception):
    def __init__(self, error_response):
        super(ChargehoundError, self).__init__(error_response['message'])
        self.status = error_response['status']
        self.message = error_response['message']


class ChargehoundAuthenticationError(ChargehoundError):
    pass


class ChargehoundBadRequestError(ChargehoundError):
    pass

class ChargehoundTimeoutError(Exception):
    pass

def create_chargehound_error(error_response):
    error = error_response['error']
    if error['status'] == 401:
        raise ChargehoundAuthenticationError(error)
    elif error['status'] == 403:
        raise ChargehoundAuthenticationError(error)
    elif error['status'] == 400:
        raise ChargehoundBadRequestError(error)
    else:
        raise ChargehoundError(error)

def create_timeout_error():
    raise ChargehoundTimeoutError('Connection timed out')


chargehound/resources.py
from chargehound.api_requestor import APIRequestor

requestor = APIRequestor()


class Disputes(object):
    """
    Create a dispute
    This method will return the created dispute.

    :return: Dispute
    """
    @classmethod
    def create(klass, dispute_id, **kwargs):
        return requestor.request('post', 'disputes',
                                 data=kwargs)

    """
    Retrieve a dispute
    This method will return a single dispute.

    :param str dispute_id: A dispute id (required)
    :return: Dispute
    """
    @classmethod
    def retrieve(klass, dispute_id):
        return requestor.request('get', 'disputes/{0}'.format(dispute_id))

    """
    A list of disputes
    This method will list all the disputes that we have synced from Stripe.
    By default the disputes will be ordered by `created`
    with the most recent dispute first.
    `has_more` will be `true` if more results are available.
    :return: Disputes
    """
    @classmethod
    def list(klass, **kwargs):
        list_params = kwargs

        return requestor.request('get', 'disputes',
                                 params=list_params)

    """
    Submitting a dispute
    You will want to submit the dispute through Chargehound after you recieve
    a notification from Stripe of a new dispute.
    With one `POST` request you can update a dispute with the
    evidence fields and send the generated response to Stripe.
    The response will have a `201` status if the submit was successful.
    The dispute will also be in the submitted state.

    :param str dispute_id: A dispute id (required)
    :return: Dispute
    """
    @classmethod
    def submit(klass, dispute_id, **kwargs):
        update = kwargs

        return requestor.request('post',
                                 'disputes/{0}/submit'.format(dispute_id),
                                 data=update)

    """
    Updating a dispute
    You can update the template and the fields on a dispute.

    :param str dispute_id: A dispute id (required)
    :return: Dispute
    """
    @classmethod
    def update(klass, dispute_id, **kwargs):
        update = kwargs

        return requestor.request('post', 'disputes/{0}'.format(dispute_id),
                                 data=update)


chargehound/api_requestor.py
from __future__ import unicode_literals

import chargehound
import requests
from chargehound.error import create_chargehound_error
from chargehound.version import VERSION
from chargehound.models import (
    ChargehoundObject, List, Dispute, Product, PastPayment, CorrespondenceItem, Response, HTTPResponse
)


class APIRequestor(object):
    def convert(self, obj):
        if obj.get('object') == 'dispute':
            obj['products'] = [Product(item) for item in obj.get('products', [])]
            obj['past_payments'] = [PastPayment(item) for item in obj.get('past_payments', [])]
            obj['correspondence'] = [CorrespondenceItem(item) for item in obj.get('correspondence', [])]
            return Dispute(obj)
        elif obj.get('object') == 'list':
            obj['data'] = [self.convert(item) for item in obj['data']]
            return List(obj)
        elif obj.get('object') == 'response':
            return Response(obj)
        else:
            return ChargehoundObject(obj)

    def parse_response(self, response):
        payload = response.json()
        if response.status_code < 400:
            model = self.convert(payload)
            setattr(model, 'response', HTTPResponse(response.status_code))
            return model
        else:
            raise create_chargehound_error(payload)

    def get_url(self, path):
        url = chargehound.protocol + chargehound.host
        if chargehound.port:
            url += ':' + chargehound.port
        url += chargehound.base_path + path
        return url

    def make_request(self, method, path, params=None, data=None,
                     callback=None):
        headers = {
            'accept': 'application/json',
            'user-agent': 'Chargehound/v1 PythonBindings/%s' % VERSION
        }

        if chargehound.version:
            headers['chargehound-version'] = chargehound.version

        auth = (chargehound.api_key or '', '')

        if method == 'get':
            return self.parse_response(requests.get(self.get_url(path),
                                       auth=auth,
                                       params=params,
                                       headers=headers,
                                       timeout=chargehound.timeout))
        elif method == 'post':
            return self.parse_response(requests.post(self.get_url(path),
                                       auth=auth,
                                       json=data,
                                       headers=headers,
                                       timeout=chargehound.timeout))

    def request(self, method, path, params=None, data=None, callback=None):
        if callback is None:
            return self.make_request(method, path, params, data)
        else:
            return self.make_request(method, path, params, data, callback)


example/submit_dispute.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# import Chargehound and set your API key

import os
import chargehound

chargehound.api_key = os.environ.get('CHARGEHOUND_API_KEY')

# List currently active disputes and then submit
# the most recent with the 'crowdfunding' template
# and update the `customer_ip` evidence field.

disputes = chargehound.Disputes.list()
first = disputes['data'][0]

products = [{'name': 'Saxophone',
             'description': 'Alto saxophone, with carrying case',
             'image': 'http://s3.amazonaws.com/chargehound/saxophone.png',
             'sku': '17283001272',
             'quantity': 1,
             'amount': 20000,
             'url': 'http://www.example.com', },
            {'name': 'Milk',
             'description': 'Semi-skimmed Organic',
             'image': 'http://s3.amazonaws.com/chargehound/milk.png',
             'sku': '26377382910',
             'quantity': '64oz',
             'amount': 400,
             'url': 'http://www.example.com'}]

submitted = chargehound.Disputes.submit(first['id'],
                                        template='crowdfunding',
                                        fields={'customer_ip': '0.0.0.0'},
                                        products=products)

print('Submitted with fields: {0}'.format(submitted['fields']))
